fix #307841: crashes and more with hbox within vbox

MarcSabatella opened the pull request:
Resolves: https://musescore.org/en/node/307841

When adding an hbox within a vbox,
we never set its tick,
and as a result other operations that depend on the tick
(like triggering layout) do not work correctly.
This is fixed by overriding HBox::tick() to check the parent.
It was also necessary to alter MeasureBase::triggerLayout(),
as it actually did nothing for hbox-within-vbox,
because it checks for prev() and next() which are nullptr here.
That check is needed to avoid triggering layout for measurebases
that are not yet added to the score.
This problem is solved here by checking the parent,
just as is done for the tick itself.

Spire42 commented on libmscore/box.h:
```diff
@@ -110,6 +110,7 @@ class HBox final : public Box {
       HBox* clone() const override      { return new HBox(*this); }
       ElementType type() const override { return ElementType::HBOX; }
 
+      Fraction tick() const;
```
This is an override of `Element::tick()`, isn't it? So it should be declared as:

    Fraction tick() const override;

I noticed that the declaration of `MeasureBase::tick()` is also missing the same `override` specifier. Maybe throw that one in too?

MarcSabatella replied:
Indeed, that's what I copied/pasted from :-).  Yes, I will add the override to both.

Spire42 commented on libmscore/measurebase.cpp:
```diff
@@ -307,8 +307,11 @@ void MeasureBase::layout()
-      if (prev() || next()) // avoid triggering layout before getting added to a score
-            score()->setLayout(tick(), -1, this);
+      // for measurebases within other measurebases (e.g., hbox within vbox), use parent
+      const MeasureBase* mb = parent() && parent()->isMeasureBase() ? toMeasureBase(parent()) : this;
```
Please add a `const` qualifier and parentheses:

    const MeasureBase* const mb = (parent() && parent()->isMeasureBase()) ? toMeasureBase(parent()) : this;

dmitrio95 commented on libmscore/box.cpp:
```diff
@@ -443,6 +443,18 @@ HBox::HBox(Score* score)
+      if (parent() && parent()->isMeasureBase())
+            return parent()->tick();
+      else
+            return MeasureBase::tick();
```
If this is intended to be an override of `Element::tick()` function (as discussed [above](https://github.com/musescore/MuseScore/pull/6339#discussion_r455359445)) then its is better to call `Box::tick()` here as `Box` is the most direct base class of `HBox`. Right now it won't change anything but it will be better in case `Box` eventually gets its own `tick()` implementation.

MarcSabatella replied:
Very good point.  I miss a clean way to say "pass the call up to parent"

dmitrio95 commented on libmscore/measurebase.cpp:
```diff
@@ -307,8 +307,11 @@ void MeasureBase::layout()
-      if (prev() || next()) // avoid triggering layout before getting added to a score
-            score()->setLayout(tick(), -1, this);
+      // for measurebases within other measurebases (e.g., hbox within vbox), use parent
+      const MeasureBase* mb = parent() && parent()->isMeasureBase() ? toMeasureBase(parent()) : this;
```
Is it possible to have boxes nested to more than one level? When trying to figure this out I only managed to crash MuseScore (see [this issue](https://musescore.org/en/node/307929)) but if there isn't any limitation on a number of nested boxes then it might be better to find a topmost `MeasureBase` parent here with a loop.

MarcSabatella replied:
It shouldn't be possible to have done this at all as far as I know, but OK, yeah, better safe than sorry.  So I should probably factor this loop out into its own function for reuse.

MarcSabatella commented:
I have restructured the code a bit to address the comments here.  Same basic idea, except now I actually do all my overrides in MeasureBase, as it turns out this simplifies things.  I have added a new member function top() that loops up the partner chain until we find the top level measurebase.  It's normally the original of course, but in the case of an hbox within a box, it's the box.  If we ever support more complex nesting structures, this should continue to work.

I also added a fix for the related issue @dmitrio95 discovered, where attempting to insert a new box with the nest box selected causes problems.  It turns out we weren't actually trying to create another nest level, but we were trying to insert the new box in from of the box in the linked list of measurebases, but the box isn't part of that list.  Fix was to use my new top() function within insertMeasure.  There's a decent chance there is some other code path where you can trigger a similar problem, and if we ever discover one, the same fix one line fix could probably be applied there too.

BTW, I do not like needing to use the const_cast, but couldn't find any good alternative.

Jojo-Schmitz commented:
Looks like it crashes in several mtests

MarcSabatella commented:
Ah yes, probably I need to check for null before calling top() in insertMeausre().  Will fix later today.

MarcSabatella commented:
That was it indeed, fixed now.
